src/allocator.rs

```diff
@@ -74,12 +102,12 @@ impl Allocator {
 
     pub fn new_atom(&mut self, v: &[u8]) -> Result<NodePtr, EvalErr> {
         let start = self.u8_vec.len() as u32;
-        if ((u32::MAX - start) as usize) < v.len() {
+        if (self.heap_limit - start as usize) < v.len() {
             return err(self.null(), "out of memory");
         }
         self.u8_vec.extend_from_slice(v);
         let end = self.u8_vec.len() as u32;
-        if self.atom_vec.len() == i32::MAX as usize {
+        if self.atom_vec.len() == self.atom_limit {
```

Comment:
Not for now necessarily, but there'd be no reason to have a separate atom limit (unless you want it for other kinds of profiling) if empty atoms all pointed to the same empty string space like in coff and elf string tables.  there'd be no way to create a new atom without taking heap_limit unless it was a clone of the empty atom which would be free.

Reply:
yeah, that's a good point. I was thinking about that when working on this patch too. We don't implement this check yet though, to ensure that all empty atom outputs reuse the one canonical null atom. Same thing could be done for "1".